utils/scale-test

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
+#
+# -*- python -*-
+#
+# Runs a .gyb scale-testing file repeatedly through swiftc while varying a
+# scaling variable 'N', collects json stats from the compiler, transforms the
+# problem to log-space and runs a linear regression to estimate the exponent on
+# the stat's growth curve relative to N.
+#
+# The estimate will be more accurate as N increases, so if you get a
+# not-terribly-convincing estimate, try increasing --begin and --end to larger
+# values.
+#
+
+import gyb, os, os.path
+
+
+def write_input_file(args, ast, d, n):
+    ifile = os.path.join(d, "in%d.swift" % n)
+    with open(ifile,'w+') as f:
+        f.write(gyb.execute_template(ast, '', N=n))
+    return ifile
+
+
+def run_once(args, ast, rng):
+    import sys, shutil, tempfile, subprocess, json
+    try:
+        d = tempfile.mkdtemp()
+        stats = os.path.join(d, "stats.json")
+        inputs = [write_input_file(args, ast, d, i) for i in rng]
+        primary = inputs[-1]
+        ofile = os.path.join(d, "out.o")
+        mode = "-c"
+        if args.parse:
+            mode = "-parse"
+        subprocess.check_call([args.swiftc_binary,
+                               "-frontend", mode,
+                               "-o", ofile,
+                               "-Xllvm", "-stats",
+                               "-Xllvm", "-stats-json",
+                               "-Xllvm", "-info-output-file=" + stats,
+                               "-primary-file", primary] + inputs)
+        with open(stats) as f:
+            return json.load(f)
+    finally:
+        shutil.rmtree(d)
+
+
+def run_many(args):
+    ast = gyb.parse_template(args.file.name, args.file.read())
+    rng = range(args.begin, args.end, args.step)
+    if args.multi_file:
+        return (rng, [run_once(args, ast, rng[0:i+1]) for i in range(len(rng))])
+    else:
+        return (rng, [run_once(args, ast, [r]) for r in rng])
+
+
+def report(args, rng, runs):
+    import numpy as np
+    bad = False
+    keys = set.intersection(*[set(j.keys()) for j in runs])
+    A = np.vstack([np.log(rng), np.ones(len(rng))]).T
+    rows = []
+    for k in keys:
+        vals = [r[k] for r in runs]
+        bounded = [max(v, 1) for v in vals]
+        b, a = np.linalg.lstsq(A, np.log(bounded))[0]
+        b = abs(b)
```

Comment:
Amusingly this doesn't quite work as written since 0.0 == -0.0 in floating point world. Moreover it turns out we were dealing with small subnormals near negative zero, not negative zero itself, so I used np.isclose and called it a day.

But I learned about python ternary today, so victory all around!